src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -219,6 +217,22 @@ public void clusterChanged(ClusterChangedEvent event) {
 
             // check and write changes in indices
             for (IndexMetaData indexMetaData : newMetaData) {
+
+                boolean shardsAllocatedOnThisNodeInLastClusterState = true;
+                if (isDataOnlyNode(state)) {
+                    boolean shardsCurrentlyAllocatedOnThisNode = shardsAllocatedOnLocalNode(state, indexMetaData);
+                    shardsAllocatedOnThisNodeInLastClusterState = shardsAllocatedOnLocalNode(event.previousState(), indexMetaData);
+
+                    if (shardsCurrentlyAllocatedOnThisNode == false) {
+                        // remove the index state for this index if it is only a data node
+                        // only delete if the last shard was removed
+                        if (shardsAllocatedOnThisNodeInLastClusterState) {
+                            removeIndexState(indexMetaData);
```

Comment:
I am not sure we need this removal logic, we already have that in `IndicesClusterStateService#applyCleanedIndices`, where an index is removed is no shards are around any more, its nice that it would be in a single place?